# MAGIC Gamma Telescope - TPOT Classification Study

The below gives information about the data set:

The data are MC generated (see below) to simulate registration of high energy gamma particles in a ground-based atmospheric Cherenkov gamma telescope using the imaging technique. Cherenkov gamma telescope observes high energy gamma rays, taking advantage of the radiation emitted by charged particles produced inside the electromagnetic showers initiated by the gammas, and developing in the atmosphere. This Cherenkov radiation (of visible to UV wavelengths) leaks through the atmosphere and gets recorded in the detector, allowing reconstruction of the shower parameters. The available information consists of pulses left by the incoming Cherenkov photons on the photomultiplier tubes, arranged in a plane, the camera. Depending on the energy of the primary gamma, a total of few hundreds to some 10000 Cherenkov photons get collected, in patterns (called the shower image), allowing to discriminate statistically those caused by primary gammas (signal) from the images of hadronic showers initiated by cosmic rays in the upper atmosphere (background). 

Typically, the image of a shower after some pre-processing is an elongated cluster. Its long axis is oriented towards the camera center if the shower axis is parallel to the telescope's optical axis, i.e. if the telescope axis is directed towards a point source. A principal component analysis is performed in the camera plane, which results in a correlation axis and defines an ellipse. If the depositions were distributed as a bivariate Gaussian, this would be an equidensity ellipse. The characteristic parameters of this ellipse (often called Hillas parameters) are among the image parameters that can be used for discrimination. The energy depositions are typically asymmetric along the major axis, and this asymmetry can also be used in discrimination. There are, in addition, further discriminating characteristics, like the extent of the cluster in the image plane, or the total sum of depositions. 

https://archive.ics.uci.edu/ml/datasets/MAGIC+Gamma+Telescope


Attribute Information:

1. fLength: continuous # major axis of ellipse [mm] 
2. fWidth: continuous # minor axis of ellipse [mm] 
3. fSize: continuous # 10-log of sum of content of all pixels [in #phot] 
4. fConc: continuous # ratio of sum of two highest pixels over fSize [ratio] 
5. fConc1: continuous # ratio of highest pixel over fSize [ratio] 
6. fAsym: continuous # distance from highest pixel to center, projected onto major axis [mm] 
7. fM3Long: continuous # 3rd root of third moment along major axis [mm] 
8. fM3Trans: continuous # 3rd root of third moment along minor axis [mm] 
9. fAlpha: continuous # angle of major axis with vector to origin [deg] 
10. fDist: continuous # distance from origin to center of ellipse [mm] 
11. class: g,h # gamma (signal), hadron (background) 

g = gamma (signal): 12332 
h = hadron (background): 6688 

For technical reasons, the number of h events is underestimated. In the real data, the h class represents the majority of the events. 

The simple classification accuracy is not meaningful for this data, since classifying a background event as signal is worse than classifying a signal event as background. For comparison of different classifiers an ROC curve has to be used. The relevant points on this curve are those, where the probability of accepting a background event as signal is below one of the following thresholds: 0.01, 0.02, 0.05, 0.1, 0.2 depending on the required quality of the sample of the accepted events for different experiments.

In [1]:
# Import required libraries
from tpot import TPOTClassifier
from sklearn.model_selection import train_test_split
import pandas as pd 
import numpy as np

In [2]:
#Load the data
telescope=pd.read_csv('MAGIC Gamma Telescope Data.csv')
telescope.head(5)

Flength    Fwidth   Fsize   Fconc  Fconc1     Fasym  Fm3long  Fm3trans  \
0   28.7967   16.0021  2.6449  0.3918  0.1982   27.7004  22.0110   -8.2027   
1   31.6036   11.7235  2.5185  0.5303  0.3773   26.2722  23.8238   -9.9574   
2  162.0520  136.0310  4.0612  0.0374  0.0187  116.7410 -64.8580  -45.2160   
3   23.8172    9.5728  2.3385  0.6147  0.3922   27.2107  -6.4633   -7.1513   
4   75.1362   30.9205  3.1611  0.3168  0.1832   -5.5277  28.5525   21.8393   

    Falpha     Fdist Class  
0  40.0920   81.8828     g  
1   6.3609  205.2610     g  
2  76.9600  256.7880     g  
3  10.4490  116.7370     g  
4   4.6480  356.4620     g

As can be seen in the above the class object is organized here, and hence for better results, I start with randomly shuffling the data.

In [3]:
telescope_shuffle=telescope.iloc[np.random.permutation(len(telescope))]
telescope_shuffle.head()

Flength   Fwidth   Fsize   Fconc  Fconc1    Fasym  Fm3long  Fm3trans  \
8677   77.4316  14.5076  2.7846  0.3596  0.2340 -75.2684  61.5001   -9.2181   
10306  18.8069  16.4272  2.2355  0.5756  0.2994  -1.0363 -10.5534   16.8676   
14532  21.7894  11.7212  2.5670  0.5264  0.3060  38.9411  13.3850   -7.8003   
8706   87.3349  35.2968  3.8790  0.1477  0.0749 -15.6010  68.7720   24.1298   
236    71.2234  17.8914  2.9445  0.2557  0.1494  68.4368  71.1737   -9.5179   

        Falpha     Fdist Class  
8677   12.3296  266.6330     g  
10306  29.6920  224.3870     g  
14532  27.4297  243.0327     h  
8706    0.7025  311.5730     g  
236     7.6150  223.1840     g

Above also rearranges the index number and below I basically reset the Index Numbers.

In [4]:
tele=telescope_shuffle.reset_index(drop=True)
tele.head()

Flength   Fwidth   Fsize   Fconc  Fconc1    Fasym  Fm3long  Fm3trans  \
0  77.4316  14.5076  2.7846  0.3596  0.2340 -75.2684  61.5001   -9.2181   
1  18.8069  16.4272  2.2355  0.5756  0.2994  -1.0363 -10.5534   16.8676   
2  21.7894  11.7212  2.5670  0.5264  0.3060  38.9411  13.3850   -7.8003   
3  87.3349  35.2968  3.8790  0.1477  0.0749 -15.6010  68.7720   24.1298   
4  71.2234  17.8914  2.9445  0.2557  0.1494  68.4368  71.1737   -9.5179   

    Falpha     Fdist Class  
0  12.3296  266.6330     g  
1  29.6920  224.3870     g  
2  27.4297  243.0327     h  
3   0.7025  311.5730     g  
4   7.6150  223.1840     g

# Pre-Processing Data

In [5]:
# Check the Data Type
tele.dtypes

Flength     float64
Fwidth      float64
Fsize       float64
Fconc       float64
Fconc1      float64
Fasym       float64
Fm3long     float64
Fm3trans    float64
Falpha      float64
Fdist       float64
Class        object
dtype: object

Class is a Categorical Variable as can be seen from above. The levels in it are:

In [6]:
for cat in ['Class']:
    print("Levels for category '{0}': {1}".format(cat, tele[cat].unique()))   

Levels for category 'Class': ['g' 'h']


Next, the categorical variable is numerically encoded since TPOT for now cannot handle categorical variables. 'g' is coded as 0 and 'h' as 1.

In [7]:
tele['Class']=tele['Class'].map({'g':0,'h':1})

A check is performed to see if there are any missing values and code then accordingly.

In [8]:
tele = tele.fillna(-999)
pd.isnull(tele).any()

Flength     False
Fwidth      False
Fsize       False
Fconc       False
Fconc1      False
Fasym       False
Fm3long     False
Fm3trans    False
Falpha      False
Fdist       False
Class       False
dtype: bool

In [9]:
tele.shape

(19020, 11)

Finally we store the class labels, which we need to predict, in a separate variable.

In [10]:
tele_class = tele['Class'].values

# Data Analysis using TPOT

To begin our analysis, we need to divide our training data into training and validation sets. The validation set is just to give us an idea of the test set error.

In [11]:
training_indices, validation_indices = training_indices, testing_indices = train_test_split(tele.index, stratify = tele_class, train_size=0.75, test_size=0.25)
training_indices.size, validation_indices.size

(14265, 4755)

After that, we proceed to calling the `fit()`, `score()` and `export()` functions on our training dataset.
An important TPOT parameter to set is the number of generations (via the `generations` kwarg). Since our aim is to just illustrate the use of TPOT, we assume the default setting of 100 generations, whilst bounding the total running time via the `max_time_mins` kwarg (which may, essentially, override the former setting). Further, we enable control for the maximum amount of time allowed for optimization of a single pipeline, via `max_eval_time_mins`.

On a standard laptop with 4GB RAM, each generation takes approximately 5 minutes to run. Thus, for the default value of 100, without the explicit duration bound, the total run time could be roughly around 8 hours.

In [12]:
tpot = TPOTClassifier(verbosity=2, max_time_mins=3, max_eval_time_mins=0.04, population_size=15)
tpot.fit(tele.drop('Class',axis=1).loc[training_indices].values, tele.loc[training_indices,'Class'].values)


Generation 1 - Current best internal CV score: 0.8399579390115667

Generation 2 - Current best internal CV score: 0.8399579390115667

Generation 3 - Current best internal CV score: 0.8447949526813879

Generation 4 - Current best internal CV score: 0.8447949526813879

Generation 5 - Current best internal CV score: 0.8464773922187172

Generation 6 - Current best internal CV score: 0.8468980021030493

Generation 7 - Current best internal CV score: 0.8476691202243252

Generation 8 - Current best internal CV score: 0.8476691202243252

3.00 minutes have elapsed. TPOT will close down.
TPOT closed during evaluation in one generation.


TPOT closed prematurely. Will use the current best pipeline.
                                                           
Best pipeline: DecisionTreeClassifier(VarianceThreshold(input_matrix, threshold=0.1), criterion=entropy, max_depth=10, min_samples_leaf=18, min_samples_split=20)


TPOTClassifier(max_eval_time_mins=0.04, max_time_mins=3, population_size=15,
               verbosity=2)

In the above, 8 generations were computed, each giving the training efficiency of fitting model on the training set. As evident, the best pipeline is the one that has the CV score of 84.767%. The model that produces this result is pipeline, consisting of a logistic regression that adds synthetic features to the input data, which then get utilized by a decision tree classifier to form the final predictions.

Next, the test error is computed for validation purposes.

In [14]:
tpot.score(tele.drop('Class',axis=1).loc[validation_indices].values, tele.loc[validation_indices, 'Class'].values)

0.8479495268138801

As can be seen, the test accuracy is 85.573%.

In [15]:
tpot.export('tpot_MAGIC_Gamma_Telescope_pipeline.py')